doc/contract-update.md

```diff
@@ -0,0 +1,56 @@
+# Upgrading and Migrating contracts
+
+We consider two approaches:  *Upgrading* existing contracts to a new version but retaining
+state data as-is, and; *migrating* users to a completely new set of contracts.
+
+We would only migrate in the event of a key loss, which prevents upgrading contracts.
+
+## Upgrade contracts approach
+
+Our contracts are already mostly divided into contracts that contain data and contracts that
+contain logic. This is fortuitous for the pattern we are going to use.
+
+`State.sol` contains most of the state, with some in `Shield.sol`.
+This division could be made complete by moving the `proposeBlock` function out of `State.sol`
+and moving the `Shield.sol` data into `State.sol` (there may be a few other minor adjustments
+TBC). `State.sol` will then only contain storage data and getters/setters for the same.
+It will also need modifying so that the addresses of the logic contracts, which it allows to
+set state, can be updated by a registered contract containing an `upgrade` function (see later).
+
+After that is done, we will never upgrade the `State.sol` contract but will upgrade the
+stateless logic contracts. The logic contracts call `State.sol` and thus we avoid the use of
+`delegateCall`. If, in the future, `State.sol` becomes insufficient for our needs, another
+data contract can be deployed alongside, which adds the new storage variables needed.
+
+We will create another contract, which proxies calls to the logic contracts. This avoids users
+having to repoint their applications to the new contracts. The `Proxy.sol` contract will also
+contain the `upgrade` function.
+
+![contract interaction](./contract-upgrade.png)
+
+### upgrade function
+
+This function will, atomically, carry out the following actions:
+
+1. Repoint `State.sol`'s registered logic contracts to the new logic contracts;
+1. Repoint `Proxy.sol` to the new contracts;
+1. Change the addresses of the contracts registered with `State.sol` to the new contract addresses;
+1. Store the address of the old contracts and the block number at which the swap-over occurred in `State.sol`; this
+will be used to help nightfall applications parse historic events and calldata. This will be in the form of
+an array, added to each time the contracts are upgraded.
+1. Call `selfdestruct` on the old logic contracts (just to return gas);
```

Comment:
Not just gas, it's actually an **extremely important thing to do.** The address of the `State` contracts are [internal to the proxy contract](https://docs.openzeppelin.com/contracts/4.x/api/proxy#Proxy-_implementation--). So using the current implementation of `Proxy.sol` there is no way of getting the contract address... And very well so. We would need to add that functionality to `State.sol`

Let's say that I manage a token "ZP" at `ZPERC20.sol`, which is proxied by some `ZProxy` contract. If this contract exposed the `ZPERC20` address, which is at some point upgraded by me to fix a vulnerability, an attacker could use the previous address to "fake" a valid version of my ERC20 contract, thus exploiting the vulnerability although the contract is not "valid" anymore. I actually saw this happening with a real-life token, and was even tempted to "save" some tokens in an BNB address, before understanding that the tokens were being managed by an old version of a contract. They were still valid ERC20 tokens, but weren't accepted by the "community". The attacker was using the old contract's tokens as a honeypot. If the contract owners used the Proxy pattern, this wouldn't be a problem.

Reply:
I never implemented the Proxy pattern, so I really don't have a hands-on knowledge on this. I'm not referring to "our" `State` contract, but the `State` contracts the Proxy patters refer to. But I may just be incredibly confused right now 😄

Reply:
Ah well, I'm sure it will all become clear once we implement it.